### PR TITLE
poll, voting: Unable to create a "protocol development" poll? Issue #2785

### LINKED ISSUE #2785: Unable to create a "protocol development" poll?
# Bug Report

**Current behavior**

I'm unable to create a poll with the "protocol development" type.

**Expected behavior**

I'm able to create such a poll.

**Steps to reproduce:**

Try to broadcast the creation of a protocol development poll.

**Gridcoin version**

5.4.8.0-leisure

**Machine specs**
- OS: Windows 10
- CPU: i7-6500u
- RAM: 24GB
- Disk Type (HD/SSD): HD

**Extra information**

"ERROR: CreateTransaction: tx size 513705 greater than the standard 100000"

I have 100k+ GRC in one address as required.

My wallet was locked, unlocked for broadcasting the poll creation transaction.

### PR DESCRIPTION
**Pull request by PrestackI**
# Implementation Proposal: Multi-Address Poll Eligibility Claims

## Problem Summary

Currently, `PollEligibilityClaim` only supports a **single address**, while `VoteWeightClaim` supports **multiple addresses**. This causes poll creation to fail when a user has 100K+ GRC spread across multiple addresses, even though they meet the balance requirement.

**Example failure case:**
- User has 120K GRC total
- Spread across 5 addresses (24K each)
- No single address has 100K → **Poll creation fails** ❌

## Proposed Solution

Replicate the "general attestation algorithm for votes" by:
1. Sorting **all UTXOs globally** by size (largest first)
2. Taking the top N UTXOs (up to `MAX_OUTPOINTS = 250`)
3. Grouping by address for signing
4. Supporting **multiple addresses** in the claim (like votes)

### CODE REVIEW

**Comment by PrestackI** [comment]
Any ideas on how to clear the OpenSUSE and Lint Check issue?

For Lint, I cleared the trailing spaces.

**Comment by jamescowens** [comment]
Go look at the lint failure log and remove the empty lines with whitespaces, or at least the whitespaces. You should run this stuff generated by AI though an editor that trims whitespaces, as AI is notorious for this.

I haven't looked at this PR in detail yet.

File: src/chainparams.cpp [hunk 1]
```diff
@@ -71,6 +71,7 @@ class CMainParams : public CChainParams {
         consensus.BlockV13Height = std::numeric_limits<int>::max();
         consensus.PollV3Height = 2671700;
         consensus.ProjectV2Height = 2671700;
+        consensus.PollMultiAddressHeight = 3950000;
```
Comment (jamescowens) [inline src/chainparams.cpp]: Normally for testing this is set at std::numeric_limits<int>::max(), and then a helper function is used to allow setting a height in the config file, similar to

```
inline bool IsSuperblockV3Enabled(int nHeight)
{
    // The argument driven override temporarily here to facilitate testing.

    return nHeight >= gArgs.GetArg("-superblockv3height", Params().GetConsensus().SuperblockV3Height);
}
```

File: src/chainparams.cpp [hunk 2]
```diff
@@ -192,6 +193,7 @@ class CTestNetParams : public CChainParams {
         consensus.BlockV13Height = 2870000;
         consensus.PollV3Height = 1944820;
         consensus.ProjectV2Height = 1944820;
+        consensus.PollMultiAddressHeight = 2870000;
```
Comment (jamescowens) [inline src/chainparams.cpp]: Same here.

**Comment by PrestackI** [comment]
Should we upgrade this anyway - even if it is not needed?

**Comment by jamescowens** [comment]
## Review: Multi-Address Poll Eligibility Claims

### 1. PollMultiAddressHeight Demarcation (Critical)

The PR does not properly gate the claim version on `PollMultiAddressHeight`. `PollClaimBuilder::BuildClaim()` in `builders.cpp` **unconditionally creates v2 claims**:

```cpp
PollEligibilityClaim claim;
claim.m_version = 2;  // Always v2, regardless of height
```

This creates a consensus fork risk:

**a) Old nodes cannot deserialize v2 claims.** The old serialization was:
```cpp
READWRITE(m_version);
READWRITE(m_address_claim);  // a single AddressClaim
```
The new serialization for v2 writes a `BalanceClaim` (a `vector<AddressClaim>`) in a different wire format. An old node seeing `m_version = 2` will still execute `READWRITE(m_address_claim)`, attempting to read a `BalanceClaim` as a single `AddressClaim`. The CompactSize prefix of the vector will be misinterpreted as the start of a public key, causing deserialization failure. **The old node rejects the block → network fork before activation.**

**b) The `BlockValidate` check is one-directional.** At `registry.cpp:1091–1099`, it only rejects v1 claims *after* activation:
```cpp
if (IsPollMultiAddressEnabled(ctx.m_pindex->nHeight)
    && payload->m_claim.m_version < 2) {
    // reject v1 after activation
}
```
There is no corresponding rule to **reject v2 claims *before* activation**. Both directions need enforcement for a clean mandatory fork.

**c) The v1 write path in serialization is broken.** In `PollEligibilityClaim::SerializationOp` (`claims.h`), the v1 write branch creates and serializes a *default-constructed empty* `AddressClaim`, not the actual claim data:
```cpp
if (m_version == 1) {
    AddressClaim single_address_claim;  // empty!
    READWRITE(single_address_claim);    // writes empty claim on serialize
    if (ser_action.ForRead()) {
        // only meaningful on read
    }
}
```
Even if the builder were changed to set `m_version = 1` before activation, the write path would produce an empty claim. The v1 write path needs to extract the single address claim from `m_balance_claim.m_address_claims[0]` before writing.

**Required fixes:**
1. `PollClaimBuilder::BuildClaim` must check `IsPollMultiAddressEnabled(nBestHeight)` and fall back to v1 single-address logic before activation.
2. `BlockValidate` should reject v2 claims *before* `PollMultiAddressHeight`.
3. The v1 write path in `SerializationOp` needs to write the actual claim data, not an empty placeholder.

---

### 2. Balance Attestation Review

**a) Duplicate outpoint / address amplification — adequately mitigated**

`BalanceClaim::WellFormed()` prevents duplicate addresses (via sorted-order check on public key) and `AddressClaim::WellFormed()` prevents duplicate outpoints within a single address. `Resolve()` verifies each output's address matches the claimed address, so cross-address outpoint duplication is structurally impossible. This is sound.

**b) `BalanceClaim::WellFormed()` accepts empty claims**

At `claims.h:160–162`:
```cpp
if (m_address_claims.empty()) {
    return true;
}
```
This means `PollEligibilityClaim::WellFormed()` passes for a poll with zero address claims. The balance check in `VerifyBalanceClaim` will catch this (total = 0 < required), but semantically an empty balance claim should not be "well-formed" for poll eligibility. The `BalanceClaim` class is shared with `VoteWeightClaim` where an empty balance claim is valid (magnitude-only voting), so this is a design tension. Consider overriding the check in `PollEligibilityClaim::WellFormed()` to require at least one address claim.

**c) Burn fee increase (probably unintentional)**

`BalanceClaim::RequiredBurnAmount()` adds **0.01 GRC per address** on top of the per-UTXO fee:
```cpp
amount += (COIN / 100) + claim.RequiredBurnAmount();
//         ^^^^^^^^^^^^   ^^^^^^^^^^^^^^^^^^^^^^^^
//         0.01/address   0.001/UTXO
```
The old `PollEligibilityClaim::RequiredBurnAmount()` was just 0.001 per UTXO. A single-address poll claim with 10 UTXOs: **old = 0.010 GRC, new = 0.020 GRC** (doubled). This per-address fee is inherited from the vote's `BalanceClaim` structure but may not be the intended fee for poll creation. If the old fee structure is desired for polls, `PollEligibilityClaim::RequiredBurnAmount()` should have its own calculation.

---

### 3. Additional Issues

**a) Null pointer dereference in `DeletePoll` (new code, high severity)**

At `registry.cpp:1245–1249`:
```cpp
PollReference* poll_ref = TryBy(payload->m_poll.m_title);
poll_ref->Notify(PollReference::PollNotificationType::POLL_DELETE);
```
`TryBy(const std::string&)` calls `TryByTitle()` which does a map lookup. The map is keyed by **lowercase** titles (set in `AddPoll` via `ToLower`), but `payload->m_poll.m_title` is in **original case**. If the title contains uppercase characters, the lookup fails, `poll_ref` is `nullptr`, and `Notify` dereferences null. Fix: use `TryBy(ToLower(payload->m_poll.m_title))` and add a null check.

**b) `PollReference::m_weight_type` not initialized in `AddPoll()` (medium severity)**

`AddPoll()` in `registry.cpp` sets `m_magnitude_weight_factor` but never sets `m_weight_type`. The field is only set in `TryReadFromDisk()`. Since `m_weight_type` default-constructs as `PollWeightType::UNKNOWN`, any code path that calls `GetActiveVoteWeight()` without first going through `TryReadFromDisk()` will see `UNKNOWN` and return `std::nullopt`. Currently `PollResult::BuildFor()` calls `TryReadFromDisk()` first, so it works in practice, but this is fragile. Add to `AddPoll()`:
```cpp
poll_ref.m_weight_type = payload->m_poll.m_weight_type.Value();
```

---

### Summary

| Severity | Issue |
|----------|-------|
| **Critical** | v2 claims produced before `PollMultiAddressHeight` will fork old nodes |
| **Critical** | `BlockValidate` doesn't reject v2 claims before activation |
| **Critical** | v1 serialization write path produces empty/broken claims |
| **High** | Null pointer dereference in `DeletePoll` notification |
| **Medium** | `m_weight_type` not set in `AddPoll()` |
| **Low** | `BalanceClaim::WellFormed()` accepts empty claims for poll eligibility |
| **Low** | Burn fee increased (probably unintentionally) for poll creation |

The fundamental issue is the lack of proper height-gated demarcation. The system needs the ability to build and validate v1 claims before activation and v2 claims at/after activation, with `BlockValidate` enforcing this **bidirectionally**.

**Comment by PrestackI** [comment]
Given that we have a workaround for this - should I proceed with modifying the pull request?

My thoughts are yes since it will bring alignment between Poll Creation and Voting mechanisms but I did see the push back to the idea in the Instant Message community stream.

Would appericate some affirmative guidance on the course of action.

**Comment by jamescowens** [comment]
Hey, how is it going? I am wondering if we should try and get this in Natasha to avoid another mandatory after v14 to fix it.

**Comment by PrestackI** [comment]
Hey, I rebased the PR and I think I addressed the issues of demarcation with commit [2d574c8](https://github.com/gridcoin-community/Gridcoin-Research/pull/2833/commits/2d574c86b3a22a81ca832cfb560a32ffee2418df). Let me know if I need to try again.
That would be great if we could get it in on Natasha. I am not familiar with testing the poll creation but I can give it a try this weekend.

**Comment by jamescowens** [comment]
I will review it tomorrow.

**Comment by PrestackI** [comment]
@jamescowens Not sure what is going on with Homebrew CI check, but it appears unrelated to this PR.

Apparently just transient issue.

**Review by jamescowens — APPROVED** [review]
Thanks for picking this back up and addressing the December review on chainparams — the `std::numeric_limits<int>::max()` default + `IsPollMultiAddressEnabled` helper with the `-pollmultiaddressheight` arg-override is exactly the pattern we've standardised on for un-coordinated future mandatories (it's also what just got asked for in #2955). The May 22 commit lineage cleanly resolves both Dec comments.

Deep-walked the rest of the change against the existing `VoteWeightClaim` precedent and the bidirectional gate-enforcement contract.

## Per-concern verdict

| # | Concern | Verdict |
|---|---|---|
| 1 | V1/v2 backward-compat split | ✓ clean — hard fork by design, serialization branch correct |
| 2 | Multi-address signing matches vote path | ✓ with intentional divergences (no `TrimSmallestOutputs`, early-exit on threshold) |
| 3 | Gate placement: build uses `nBestHeight`, validation uses `ctx.m_pindex->nHeight` | ✓ bidirectional enforcement correct |
| 4 | Replay safety — `PackPollMessage` binds poll content + nTime + tx inputs | ✓ |
| 5 | `MAX_OUTPOINTS = 250` truncation | ✓ matches the vote-path precedent; pre-existing edge case for >250-UTXO wallets |
| 6 | Test coverage | no new tests, but no voting-test infrastructure exists at all in the project — consistent with precedent (not blocking) |
| 7 | Lock annotations (`EXCLUSIVE_LOCKS_REQUIRED(cs_main)`) | ✓ honestly applied at every reachable call site |
| 8 | December review (`int_max` + helper convention) | ✓ both addressed |

## Fixup items I'll push directly via maintainer-can-modify

These are mechanical enough that splitting into back-and-forth doesn't add value:

1. **Register `-pollmultiaddressheight` in `hidden_args` (init.cpp:673-676).** Same bug class as #2822/HTLC missed. Without it the override produces an "argument not registered" warning at startup. One-line addition alongside the existing `-projectv4height` / `-superblockv3height` / `-blockv14height` / `-autogreylistauditheight` entries.

2. **`LogPrint("%u address(es)", ...size())` in builders.cpp** — `%u` against `size_t` is technically UB on 64-bit; should be `%" PRIszu "` matching sibling LogPrints in the same file (lines 110, 653, 716).

3. **Comment cleanup at `chainparams.h:169`** — `"The argument driven override temporarily here to facilitate testing."` has an extra "here"; cosmetic.

## RPC schema change worth surfacing

`src/rpc/voting.cpp:305` renames `address_claim` → `balance_claim` in `PollClaimToJson`. Silent breaking change for external tooling parsing poll-details JSON (block explorers, dashboards). Not a code blocker — but worth a release-notes mention when whichever version bundles V15 / this multi-address change ships. Happy to add a CHANGELOG line when the version cadence is decided; can wait for the v15 coordination.

## Bundled-but-unrelated improvements (acknowledged, not blocking)

Three unrelated fixes piggyback on this PR. All correct and useful:
- `PollReference::m_weight_type` in-class init + `AddPoll` initialization — aligns with the project's [in-class-init policy](https://github.com/gridcoin-community/Gridcoin-Research/blob/development/doc/developer-notes.md) we recently re-emphasised
- `DeletePoll` case-insensitive `ToLower(...)` lookup + `nullptr` guard before `Notify` — real fix for a real bug
- Early-return in `DeletePoll` when the lookup misses — analysis shows the skipped erases would be no-ops anyway (map is keyed by `ToLower`), only behavioural skip is `m_latest_poll = nullptr`, which is arguably correct when nothing was deleted

Ideally these would be split in future PRs of similar shape, but they don't change the verdict here.

## Minor pre-existing observations (not introduced by this PR — separate follow-up worthy)

- `AddressClaim::WellFormed` allows duplicate-equal outpoints (`<` check doesn't catch equal-consecutive). Would let a signer double-count one UTXO; pre-existed under v1 single-address.
- `BalanceClaim::WellFormed()` returns `true` for an empty claim — the PR correctly guards at the `PollEligibilityClaim` layer (`!m_balance_claim.m_address_claims.empty()`), so no exposure here.

## Pre-gate v1 behaviour change (minor, temporary)

New `BuildV1Claim` iterates `outputs_by_address` (a `std::map<CKeyID, ...>`) in CKeyID byte order rather than descending-balance order. Pre-gate, a wallet with multiple 100K+ addresses now picks the one with smallest CKeyID instead of the largest balance. Visible behaviour change, no consensus implication, temporary (only used pre-gate). Leaving as-is since the v1 path becomes inactive once V15 / the gate height activates.

## Bottom line: approve

`PollMultiAddressHeight` is set to `int_max` on both nets, so the consensus code is inert by default — actual deployment is a follow-up coordination matter (probably with the next mandatory cluster, V15-style). I'll push the three fixup commits above directly via maintainer-can-modify so this can land cleanly when the V15 cycle is ready.

**Comment by jamescowens** [comment]
@PrestackI — integration-testing #2833 + #2839 together on the
isolated-testnet 4-node mesh surfaced two wallet-side construction
issues for poll-creation txs that I think are worth addressing before
#2833 ships. Both fixes are wallet-only, **consensus-invariant**
(verifiers accept what they accept; how the wallet selects outpoints
doesn't matter for validation), and small enough to be clean
cherry-picks on top of your PR.

Branch: `jamescowens/Gridcoin-Research:testnet_v13_pr2833_pr2839_integration`

Two commits, in order:

- [`3076b1900`](https://github.com/jamescowens/Gridcoin-Research/commit/3076b1900) — `voting: dust-resistant v1 poll claim selection (#2833 follow-up)`
- [`5dec83185`](https://github.com/jamescowens/Gridcoin-Research/commit/5dec83185) — `voting, rpc: largest-first fee selection + optional fee_outpoint param (#2833 follow-up)`

Cherry-pick both, not merge — the integration branch carries the
`testnet_v13_checkpoint_removed` hack that must not reach `development`.

## Issue A — `BuildV1Claim` selection is not dust-resistant

In `PollClaimBuilder::BuildV1Claim` (`src/gridcoin/voting/builders.cpp`),
the pre-PR loop iterates `outputs_by_address` in keyID-lex order, picks
the first address whose `m_total_amount >= POLL_REQUIRED_BALANCE`, and
then `m_outpoints.resize(MAX_OUTPOINTS)` — keeping the **first 250
outpoints in wallet-iteration order**, irrespective of their amounts.

The threshold check uses the full address balance, but the claim only
carries those first 250 outpoints. On an address whose balance is spread
across many low-value UTXOs (heavily-staked researcher addresses, dust-
fragmented exchange hot wallets), the included outpoints may sum to far
less than 100K even though the address total exceeds it. The verifier
(`registry.cpp:229` sums the included outpoints) would then reject the
tx — a wallet-side construction that cannot be validated.

Commit `3076b1900` rewrites the selection:

1. Up-front `m_wallet.GetBalance() < POLL_REQUIRED_BALANCE` check for
   clearer UX on under-funded wallets (also added to the v2 path for
   symmetry).
2. Group outputs by address using `AddressOutputs::Add()` so both
   `m_outpoints` and `m_amounts` are populated in lockstep.
3. Filter to candidate addresses (per-address total ≥ threshold).
4. Sort candidates by total balance **descending**.
5. For each candidate, sort its outpoints by amount **descending** and
   walk top-down, accumulating until either the running sum reaches
   the threshold (success — build the claim) or the outpoint count hits
   `MAX_OUTPOINTS` (move to next candidate).
6. If no candidate succeeds, throw with a hint about UTXO
   consolidation.

The included outpoints now sum to ≥ 100K **by construction**.

The v2 path (`BuildClaim` post-gate) was already dust-resistant by
construction (global descending sort + early-exit at threshold), so it's
unchanged except for the upfront balance check.

## Issue B — `SelectFinalInputs` produces oversized contract txs on dust-heavy wallets

`SelectFinalInputs<PollPayload>` calls `wallet.CreateTransaction(...)`
with `coinControl=nullptr`. `CWallet::SelectCoins` then falls through to
`SelectSmallestCoins` for contract txs — a heuristic that prefers the
smallest UTXOs first to preserve large coins for staking. On a wallet
with extensive dust, gathering 50 GRC of fee payment from dust selects
hundreds or thousands of small inputs, producing a tx that exceeds the
100 KB standard tx size limit. We observed this empirically on the
isolated testnet:

```
ERROR: CreateTransaction: tx size 505271 greater than standard 100000
```

That's a 505 **kilobyte** poll-creation tx, rejected by the standard
size policy. The user sees `addpoll → "Could not create transaction.
See debug.log."` and is dead in the water.

Commit `5dec83185` introduces a largest-first fee-selection policy for
contract txs (poll-creation **and** vote-creation — both go through
`SelectFinalInputs`):

1. New `PickLargestFirstFeeInputs()` helper sorts available coins by
   amount descending and accumulates outpoints top-down via
   `CCoinControl::Select` until burn_fee + 1 GRC headroom is covered.
2. `SelectFinalInputs<PayloadType>` accepts an optional explicit
   `COutPoint` parameter. When provided (user coin-control), the
   wallet pins exactly that outpoint. When omitted, the largest-first
   default fires.
3. `CCoinControl::HasSelected()` forces `CWallet::SelectCoins` to use
   ONLY the pre-selected outpoints, bypassing `SelectSmallestCoins`
   entirely.
4. `PollBuilder::SetFeeOutpoint(const COutPoint&)` fluent setter
   threads the optional outpoint from the RPC layer to
   `BuildContractTx` to `SelectFinalInputs`.
5. `addpoll` RPC gains an optional 10th parameter `fee_outpoint` in
   `"<txid>:<vout>"` form, with minimal format validation. The wallet
   reports `Insufficient funds.` via the existing path if the pinned
   outpoint is unavailable.

## Empirical validation on the integration mesh

Test wallet: instance 8, 372,791 txs, balance 24.27 M GRC, 159
addresses with 3 of them ≥ 100 K. The dominant address holds 22 M GRC
from years of stake rewards.

| Attribute | Pre-fix (PR #2833 at `d12b63994`) | Post-fix (`5dec83185`) |
|---|---|---|
| Poll-creation tx size | 505,271 bytes — **rejected** | **508 bytes** — accepted |
| vin count | thousands of small UTXOs | **1** large-value UTXO |
| Result | `tx size > 100000` error | Confirmed in block, 4+ confirmations |
| **Reduction** | — | **~1000×** smaller tx |

The `fee_outpoint` parameter test: pinned a deliberately non-largest
UTXO (806.36 GRC, well outside the wallet's top-5) on instance 9,
called `addpoll` with `fee_outpoint = "<txid>:1"`, and verified the
resulting tx vin contains exactly that outpoint and nothing else. Tx
propagated to all 4 mesh nodes, confirmed at the same block hash, and
appeared in `listpolls` on every node — clean coin-control with full
consensus invariance.

## Consensus invariance is what makes the cherry-pick attractive

Both fixes are wallet-side selection policy. The on-chain claim format
is identical to what #2833 already ships; the verifier path
(`registry.cpp` summing included outpoints, checking signatures) is
unchanged. A node running pre-fix code accepts a tx from a post-fix
wallet, and vice versa. So no consensus coordination, no version gate,
no field bump. The fixes can land alongside #2833 (cherry-picked) or
as an immediate follow-up PR after #2833 merges — either way, no
deployment timing concern.

## Phase A test status

- ✅ A.4.1 — fragmented wallet v1 fails with clear "no single address" error (correct behavior, dust-resistant builder routes there for genuinely fragmented wallets)
- ✅ A.2.1 — heavy wallet v1 succeeds with 508-byte tx (previously rejected; this is the headline fix)
- ✅ `fee_outpoint` param exercise — pinned UTXO honored exactly, tx propagated and confirmed
- ⏳ A.2.2 — heavy wallet v2 (post-gate) — gate at 2,775,683, currently ~37 blocks away; **testing in progress**, will report when the gate activates
- ⏳ A.4.2 — fragmented wallet v2 (post-gate, multi-address) — same
- ⏳ A.4.3 — multi-node v2 consensus across mesh — same

I'll update this thread once the chain crosses the gate and the v2
tests complete (~hour, depending on testnet stake-mining rate). Given
the v2 path is already dust-resistant by construction and consensus-
invariance is empirically demonstrated for v1, my expectation is the
v2 results will be clean too — but happy to report whatever we see.

## Suggested action

Cherry-pick `3076b1900` and `5dec83185` onto #2833. Net diff:

- `src/gridcoin/voting/builders.h`: +23 lines (SetFeeOutpoint declaration + member)
- `src/gridcoin/voting/builders.cpp`: +212 / -24 lines (v1 rewrite + helper + SelectFinalInputs change)
- `src/rpc/voting.cpp`: +37 lines (addpoll fee_outpoint param + parsing)

No on-disk format change, no consensus rule change, no nWalletVersion
bump. Old wallets remain interoperable with new ones; the fix makes
poll creation **possible** on dust-fragmented wallets where it
previously was effectively prohibited.

Happy to land these as a separate follow-up PR if you'd rather keep
#2833 scope tight. Or to take any review feedback before/during
cherry-pick.

**Comment by jamescowens** [comment]
Post-gate Phase A complete. All tests pass.

## Final Phase A matrix

| Test | Status | Tx size | Notes |
|---|---|---|---|
| A.4.1 — fragmented v1 must fail | ✅ | n/a | "No single address has 100K GRC" — correct (dust-resistant v1 routes here for genuinely fragmented wallets) |
| A.2.1 — heavy wallet v1 succeeds | ✅ | **508 B** | (was 505,271 B pre-fix; **994× reduction**) |
| `fee_outpoint` param exercise | ✅ | 1,258 B | Pinned UTXO honored exactly across 4-node propagation |
| A.2.2 — heavy wallet v2 (post-gate) | ✅ | **603 B** | v2 path active, 1 vin from largest-first selection |
| A.4.2 — fragmented v2 multi-address (post-gate) | ✅ | **9,861 B** | Canonical multi-address case: 250 outpoints across 5 addresses, claim envelope ~9.4 KB |
| A.4.3 — multi-node consensus on v2 | ✅ | same | Both polls in same block `8ba995fee4afd941…` on all 4 nodes; `listpolls` count = 16 on every node |

## A.4.2 walkthrough — the canonical "no single address has 100K" case

Instance 11's wallet shape (deliberately fragmented for this test):

```
Address 1:  50 UTXOs × 411 GRC = 20,550 GRC
Address 2:  50 UTXOs × 406 GRC = 20,300 GRC
Address 3:  50 UTXOs × 401 GRC = 20,050 GRC
Address 4:  50 UTXOs × 396 GRC = 19,800 GRC
Address 5: 100 UTXOs × 391 GRC = 39,100 GRC
                            Total = 119,800 GRC (no single address ≥ 100K)
```

v1 (pre-gate) → correctly fails (A.4.1).
v2 (post-gate) → v2 builder's global descending sort walks top-down:

- 50 UTXOs from Addr 1 (411 ea) → sum 20,550
- 50 UTXOs from Addr 2 (406 ea) → sum 40,850
- 50 UTXOs from Addr 3 (401 ea) → sum 60,900
- 50 UTXOs from Addr 4 (396 ea) → sum 80,700
- 50 UTXOs from Addr 5 (391 ea) → sum 100,250 → **early-exit at threshold**

Result: 250 outpoints used, all 5 addresses contribute. `m_balance_claim.m_address_claims.size() = 5`. 9,861-byte tx (well below the 100 KB standard limit). Confirmed in block `8ba995fee4afd941…` on all 4 nodes.

This is the headline scenario PR #2833 is designed to enable. Confirmed working end-to-end with the two follow-up commits applied.

## Both fixes consensus-invariant — verified

Two distinct claim shapes (v1 single-address, v2 multi-address) and one explicit-fee-outpoint claim, all built by the new wallet code, all accepted by every node in the mesh, all confirmed at identical block placements. The pre-fix and post-fix verifier paths produce identical accept/reject outcomes — wallets just no longer construct invalid or oversized txs in scenarios that previously stuck them.

## Final recommendation, unchanged

Cherry-pick [`3076b1900`](https://github.com/jamescowens/Gridcoin-Research/commit/3076b1900) and [`5dec83185`](https://github.com/jamescowens/Gridcoin-Research/commit/5dec83185) onto #2833. Net diff:

- `src/gridcoin/voting/builders.h`: +23 lines
- `src/gridcoin/voting/builders.cpp`: +212 / -24 lines
- `src/rpc/voting.cpp`: +37 lines

No on-disk format change, no consensus rule change, no nWalletVersion bump. Wallet-side construction policy only.

Happy to land these as a separate follow-up PR after #2833 merges if you'd rather keep #2833 scope tight. Either way, the empirical case is now closed.

**Comment by jamescowens** [comment]
@PrestackI — merging this PR as-is. The two follow-up commits I posted earlier (`3076b1900` + `5dec83185`) will become a **separate small PR** opened against fresh `development` after this merges. Rationale: the dust-resistance + largest-first fee + new `fee_outpoint` RPC parameter together amount to a feature-grade change that deserves its own review thread, its own CI cycle, and a clean description — not a stealth-add to your already-approved PR. The Phase A test matrix and the 994× tx-size measurement from the integration mesh will all be referenced from the new PR's description.

Thanks for an excellent piece of work on this one. Multi-address poll eligibility is the kind of contract-payload upgrade that closes a real usability gap (researchers and exchanges with balance fragmented across addresses now have a path), and the v2 envelope design is clean. Merging now.
